File: src/backend/app/schemas/library.py
"""
Pydantic schema models for library data validation, serialization, and API responses.

This module defines the data models used for library-related operations in the
Molecular Data Management and CRO Integration Platform, with appropriate validation
rules, type annotations, and documentation.
"""

from pydantic import BaseModel, Field, validator  # version 2.0+
from typing import Optional, List, Dict, Any, Union  # standard library
from uuid import UUID  # standard library
from datetime import datetime  # standard library

from ..models.library import Library
from .user import UserResponse
from .molecule import MoleculeResponse
from ..utils.validation_utils import validate_string_length, validate_library_name, validate_description


class LibraryBase(BaseModel):
    """Base Pydantic model for library data with common fields."""
    name: str = Field(
        ..., 
        description="Name of the library",
        example="High Activity Compounds",
        min_length=1,
        max_length=100
    )
    description: Optional[str] = Field(
        None, 
        description="Detailed description of the library's purpose",
        example="Collection of compounds with high binding affinity to target protein A"
    )

    @validator('name')
    def validate_library_name(cls, v):
        """Validates that a library name is properly formatted and not empty."""
        if v is None or v.strip() == "":
            raise ValueError("Library name cannot be empty")
        
        if not validate_string_length(v, 1, 100):
            raise ValueError("Library name must be between 1 and 100 characters")
        
        if not validate_library_name(v):
            raise ValueError("Library name contains invalid characters")
        
        return v
    
    @validator('description')
    def validate_description(cls, v):
        """Validates that a library description is properly formatted."""
        if v is None:
            return None
        
        if not validate_description(v):
            raise ValueError("Library description contains invalid characters")
        
        return v
    
    class Config:
        from_attributes = True


class LibraryCreate(LibraryBase):
    """Pydantic model for creating a new library."""
    created_by: Optional[UUID] = Field(
        None, 
        description="ID of the user creating the library"
    )


class LibraryUpdate(BaseModel):
    """Pydantic model for updating an existing library."""
    name: Optional[str] = Field(
        None, 
        description="Updated name of the library",
        example="Optimized High Activity Compounds"
    )
    description: Optional[str] = Field(
        None, 
        description="Updated description of the library's purpose",
        example="Refined collection of high-affinity compounds"
    )

    @validator('name')
    def validate_library_name(cls, v):
        """Validates that a library name is properly formatted if provided."""
        if v is None:
            return None
        
        if v.strip() == "":
            raise ValueError("Library name cannot be empty")
        
        if not validate_string_length(v, 1, 100):
            raise ValueError("Library name must be between 1 and 100 characters")
        
        if not validate_library_name(v):
            raise ValueError("Library name contains invalid characters")
        
        return v
    
    @validator('description')
    def validate_description(cls, v):
        """Validates that a library description is properly formatted if provided."""
        if v is None:
            return None
        
        if not validate_description(v):
            raise ValueError("Library description contains invalid characters")
        
        return v
    
    class Config:
        from_attributes = True


class LibraryInDBBase(LibraryBase):
    """Base Pydantic model for library data as stored in the database."""
    id: UUID = Field(..., description="Unique identifier for the library")
    created_by: UUID = Field(..., description="ID of the user who created this library")
    created_at: datetime = Field(..., description="Timestamp when the library was created")
    updated_at: Optional[datetime] = Field(None, description="Timestamp when the library was last updated")

    class Config:
        from_attributes = True


class LibraryResponse(LibraryInDBBase):
    """Pydantic model for library data in API responses."""
    creator: Optional[UserResponse] = Field(
        None, 
        description="User who created this library"
    )
    molecule_count: int = Field(
        0, 
        description="Number of molecules in this library",
        ge=0
    )

    class Config:
        from_attributes = True


class LibraryDetailResponse(LibraryInDBBase):
    """Pydantic model for detailed library data in API responses, including molecules."""
    creator: Optional[UserResponse] = Field(
        None, 
        description="User who created this library"
    )
    molecules: List[MoleculeResponse] = Field(
        default_factory=list,
        description="List of molecules in this library"
    )

    class Config:
        from_attributes = True


class LibraryListResponse(BaseModel):
    """Pydantic model for paginated list of libraries in API responses."""
    items: List[LibraryResponse] = Field(..., description="List of libraries")
    total: int = Field(..., description="Total number of libraries matching the criteria")
    page: int = Field(..., description="Current page number")
    size: int = Field(..., description="Number of items per page")


class LibraryFilter(BaseModel):
    """Pydantic model for filtering libraries by various criteria."""
    name: Optional[str] = Field(
        None, 
        description="Filter by library name (partial match)",
        example="activity"
    )
    created_by: Optional[UUID] = Field(
        None, 
        description="Filter by creator user ID"
    )
    created_after: Optional[datetime] = Field(
        None, 
        description="Filter by creation date (after)"
    )
    created_before: Optional[datetime] = Field(
        None, 
        description="Filter by creation date (before)"
    )
    contains_molecule_id: Optional[UUID] = Field(
        None, 
        description="Filter libraries containing a specific molecule"
    )
    sort_by: Optional[str] = Field(
        None, 
        description="Field to sort by",
        example="name"
    )
    sort_desc: Optional[bool] = Field(
        False, 
        description="Sort in descending order if true"
    )


class LibraryMoleculeOperation(BaseModel):
    """Pydantic model for adding or removing molecules from a library."""
    library_id: UUID = Field(
        ..., 
        description="ID of the library to modify"
    )
    molecule_ids: List[UUID] = Field(
        ..., 
        description="List of molecule IDs to add or remove"
    )

File: src/web/src/types/library.ts
import { UUID } from 'uuid';
import { Molecule } from './molecule';
import { User } from './user';

/**
 * Enum for library molecule operations
 * Defines operations that can be performed on molecules within a library
 */
export enum LibraryOperationType {
  /** Add molecules to a library */
  ADD = 'add',
  /** Remove molecules from a library */
  REMOVE = 'remove'
}

/**
 * Interface for library data with basic information
 * Represents a collection of molecules created by a user
 */
export interface Library {
  /** Unique identifier for the library */
  id: UUID;
  /** Name of the library */
  name: string;
  /** Optional description of the library's purpose or contents */
  description?: string;
  /** ID of the user who created this library */
  created_by: number; // Using number to match User.id type
  /** Date and time when the library was created */
  created_at: string;
  /** Date and time when the library was last updated */
  updated_at?: string;
  /** Number of molecules in this library */
  molecule_count: number;
  /** User who created this library */
  creator?: {
    id: number;
    email: string;
  };
}

/**
 * Interface for detailed library data including contained molecules
 * Extends the basic Library interface with molecule information
 */
export interface LibraryDetail extends Library {
  /** Molecules contained in this library */
  molecules: Molecule[];
}

/**
 * Interface for creating a new library
 * Contains the minimum required fields to create a library
 */
export interface LibraryCreate {
  /** Name of the library */
  name: string;
  /** Optional description of the library's purpose or contents */
  description?: string;
  /** Optional list of molecule IDs to initially include in the library */
  molecule_ids?: UUID[];
}

/**
 * Interface for updating an existing library
 * Contains fields that can be modified after creation
 */
export interface LibraryUpdate {
  /** Updated name of the library */
  name?: string;
  /** Updated description of the library */
  description?: string;
}

/**
 * Interface for filtering libraries by various criteria
 * Used for searching and organizing libraries
 */
export interface LibraryFilter {
  /** Filter by library name (partial match) */
  name?: string;
  /** Filter by creator user ID */
  created_by?: number; // Using number to match User.id type
  /** Filter by libraries created after this date */
  created_after?: string;
  /** Filter by libraries created before this date */
  created_before?: string;
  /** Filter by libraries containing a specific molecule */
  contains_molecule_id?: UUID;
  /** Page number for pagination */
  page?: number;
  /** Number of items per page */
  page_size?: number;
  /** Field to sort results by */
  sort_by?: string;
  /** Whether to sort in descending order */
  sort_desc?: boolean;
}

/**
 * Interface for paginated list of libraries in API responses
 * Used for returning large sets of library data with pagination
 */
export interface LibraryListResponse {
  /** Array of library items */
  items: Library[];
  /** Total number of libraries matching the filter */
  total: number;
  /** Current page number */
  page: number;
  /** Number of items per page */
  size: number;
}

/**
 * Interface for adding or removing molecules from a library
 * Used for batch operations on library contents
 */
export interface LibraryMoleculeOperation {
  /** Type of operation to perform (add or remove) */
  operation: LibraryOperationType;
  /** List of molecule IDs to operate on */
  molecule_ids: UUID[];
}

/**
 * Interface for library state in Redux store
 * Represents the complete state of the library module in the application
 */
export interface LibraryState {
  /** List of libraries */
  libraries: Library[];
  /** Currently selected library detail */
  currentLibrary: LibraryDetail | null;
  /** Whether library data is being loaded */
  loading: boolean;
  /** Error message if loading failed */
  error: string | null;
  /** Total number of libraries matching the current filter */
  totalLibraries: number;
  /** Current page number */
  currentPage: number;
  /** Number of items per page */
  pageSize: number;
}

/**
 * Type for supported library export formats
 * Defines the available file formats for exporting library data
 */
export type LibraryExportFormat = 'csv' | 'sdf';